File: cloud-mobile-integration/firebase/ios-swift/FirebaseAuthManager.swift
import Foundation
import FirebaseAuth
import GoogleSignIn
import AuthenticationServices
import Combine

/// Firebase Authentication Manager for SwiftUI
/// Handles user authentication with Firebase Auth
@MainActor
class FirebaseAuthManager: NSObject, ObservableObject {
    static let shared = FirebaseAuthManager()
    
    @Published var currentUser: User?
    @Published var authState: FirebaseAuthState = .unauthenticated
    
    private var authStateHandle: AuthStateDidChangeListenerHandle?
    private var cancellables = Set<AnyCancellable>()
    
    override private init() {
        super.init()
        setupAuthStateListener()
    }
    
    /// Setup auth state listener
    private func setupAuthStateListener() {
        authStateHandle = Auth.auth().addStateDidChangeListener { [weak self] _, user in
            self?.currentUser = user
            self?.authState = user != nil ? .authenticated(user!) : .unauthenticated
            print("Auth state changed: \(user?.uid ?? "nil")")
        }
        
        // Set initial user
        currentUser = Auth.auth().currentUser
        authState = currentUser != nil ? .authenticated(currentUser!) : .unauthenticated
    }
    
    deinit {
        if let handle = authStateHandle {
            Auth.auth().removeStateDidChangeListener(handle)
        }
    }
    
    /// Sign up with email and password
    func signUpWithEmail(
        email: String,
        password: String
    ) async throws -> User {
        let authResult = try await Auth.auth().createUser(
            withEmail: email,
            password: password
        )
        return authResult.user
    }
    
    /// Sign in with email and password
    func signInWithEmail(
        email: String,
        password: String
    ) async throws -> User {
        let authResult = try await Auth.auth().signIn(
            withEmail: email,
            password: password
        )
        return authResult.user
    }
    
    /// Sign in with Google
    func signInWithGoogle() async throws -> User {
        // Get the client ID from Firebase
        guard let clientID = FirebaseApp.app()?.options.clientID else {
            throw NSError(
                domain: "FirebaseAuthManager",
                code: -1,
                userInfo: [NSLocalizedDescriptionKey: "Missing Firebase client ID"]
            )
        }
        
        // Configure Google Sign-In
        let config = GIDConfiguration(clientID: clientID)
        GIDSignIn.sharedInstance.configuration = config
        
        // Get root view controller
        guard let windowScene = UIApplication.shared.connectedScenes.first as? UIWindowScene,
              let rootViewController = windowScene.windows.first?.rootViewController else {
            throw NSError(
                domain: "FirebaseAuthManager",
                code: -1,
                userInfo: [NSLocalizedDescriptionKey: "No root view controller"]
            )
        }
        
        // Start sign-in flow
        let result = try await GIDSignIn.sharedInstance.signIn(
            withPresenting: rootViewController
        )
        
        guard let idToken = result.user.idToken?.tokenString else {
            throw NSError(
                domain: "FirebaseAuthManager",
                code: -1,
                userInfo: [NSLocalizedDescriptionKey: "Missing ID token"]
            )
        }
        
        let credential = GoogleAuthProvider.credential(
            withIDToken: idToken,
            accessToken: result.user.accessToken.tokenString
        )
        
        let authResult = try await Auth.auth().signIn(with: credential)
        return authResult.user
    }
    
    /// Sign in with Apple
    func signInWithApple() async throws -> User {
        let provider = OAuthProvider(providerID: "apple.com")
        let authResult = try await Auth.auth().signIn(with: provider)
        return authResult.user
    }
    
    /// Sign in with Apple (using ASAuthorizationController)
    func signInWithAppleController(
        authorization: ASAuthorization
    ) async throws -> User {
        guard let appleIDCredential = authorization.credential as? ASAuthorizationAppleIDCredential else {
            throw NSError(
                domain: "FirebaseAuthManager",
                code: -1,
                userInfo: [NSLocalizedDescriptionKey: "Invalid Apple ID credential"]
            )
        }
        
        guard let appleIDToken = appleIDCredential.identityToken,
              let idTokenString = String(data: appleIDToken, encoding: .utf8) else {
            throw NSError(
                domain: "FirebaseAuthManager",
                code: -1,
                userInfo: [NSLocalizedDescriptionKey: "Unable to fetch identity token"]
            )
        }
        
        let credential = OAuthProvider.credential(
            withProviderID: "apple.com",
            idToken: idTokenString,
            rawNonce: appleIDCredential.authorizationCode.flatMap { String(data: $0, encoding: .utf8) }
        )
        
        let authResult = try await Auth.auth().signIn(with: credential)
        return authResult.user
    }
    
    /// Sign in with Facebook
    func signInWithFacebook(accessToken: String) async throws -> User {
        let credential = FacebookAuthProvider.credential(withAccessToken: accessToken)
        let authResult = try await Auth.auth().signIn(with: credential)
        return authResult.user
    }
    
    /// Sign in with phone number
    func signInWithPhoneNumber(
        phoneNumber: String
    ) async throws -> String {
        let verificationID = try await PhoneAuthProvider.provider().verifyPhoneNumber(
            phoneNumber,
            uiDelegate: nil
        )
        return verificationID
    }
    
    /// Verify phone code
    func verifyPhoneCode(
        verificationID: String,
        code: String
    ) async throws -> User {
        let credential = PhoneAuthProvider.provider().credential(
            withVerificationID: verificationID,
            verificationCode: code
        )
        
        let authResult = try await Auth.auth().signIn(with: credential)
        return authResult.user
    }
    
    /// Sign in anonymously
    func signInAnonymously() async throws -> User {
        let authResult = try await Auth.auth().signInAnonymously()
        return authResult.user
    }
    
    /// Sign out
    func signOut() throws {
        try Auth.auth().signOut()
        currentUser = nil
        authState = .unauthenticated
    }
    
    /// Get current user
    func getCurrentUser() -> User? {
        return Auth.auth().currentUser
    }
    
    /// Check if user is authenticated
    func isAuthenticated() -> Bool {
        return Auth.auth().currentUser != nil
    }
    
    /// Send email verification
    func sendEmailVerification() async throws {
        try await currentUser?.sendEmailVerification()
    }
    
    /// Send password reset email
    func sendPasswordResetEmail(email: String) async throws {
        try await Auth.auth().sendPasswordReset(withEmail: email)
    }
    
    /// Update email
    func updateEmail(newEmail: String) async throws {
        try await currentUser?.updateEmail(to: newEmail)
    }
    
    /// Update password
    func updatePassword(newPassword: String) async throws {
        try await currentUser?.updatePassword(to: newPassword)
    }
    
    /// Update profile
    func updateProfile(
        displayName: String? = nil,
        photoURL: URL? = nil
    ) async throws {
        let changeRequest = currentUser?.createProfileChangeRequest()
        
        if let displayName = displayName {
            changeRequest?.displayName = displayName
        }
        
        if let photoURL = photoURL {
            changeRequest?.photoURL = photoURL
        }
        
        try await changeRequest?.commitChanges()
    }
    
    /// Delete account
    func deleteAccount() async throws {
        try await currentUser?.delete()
        currentUser = nil
        authState = .unauthenticated
    }
    
    /// Get ID token
    func getIdToken(forceRefresh: Bool = false) async throws -> String {
        guard let user = currentUser else {
            throw NSError(
                domain: "FirebaseAuthManager",
                code: -1,
                userInfo: [NSLocalizedDescriptionKey: "No authenticated user"]
            )
        }
        
        return try await user.getIDToken(forcingRefresh: forceRefresh)
    }
    
    /// Link email credential
    func linkEmailCredential(
        email: String,
        password: String
    ) async throws -> User {
        let credential = EmailAuthProvider.credential(
            withEmail: email,
            password: password
        )
        
        let authResult = try await currentUser?.link(with: credential)
        return authResult!.user
    }
    
    /// Re-authenticate user
    func reauthenticate(
        email: String,
        password: String
    ) async throws {
        let credential = EmailAuthProvider.credential(
            withEmail: email,
            password: password
        )
        
        try await currentUser?.reauthenticate(with: credential)
    }
    
    /// Reload user data
    func reloadUser() async throws {
        try await currentUser?.reload()
    }
    
    /// Check if email is verified
    func isEmailVerified() -> Bool {
        return currentUser?.isEmailVerified ?? false
    }
    
    /// Get user metadata
    func getUserMetadata() -> UserMetadata? {
        return currentUser?.metadata
    }
    
    /// Unlink provider
    func unlinkProvider(_ providerID: String) async throws -> User {
        let user = try await currentUser?.unlink(fromProvider: providerID)
        return user!
    }
}

/// Firebase authentication state
enum FirebaseAuthState: Equatable {
    case authenticated(User)
    case unauthenticated
    case error(String)
    
    var isAuthenticated: Bool {
        if case .authenticated = self {
            return true
        }
        return false
    }
    
    static func == (lhs: FirebaseAuthState, rhs: FirebaseAuthState) -> Bool {
        switch (lhs, rhs) {
        case (.authenticated(let lhsUser), .authenticated(let rhsUser)):
            return lhsUser.uid == rhsUser.uid
        case (.unauthenticated, .unauthenticated):
            return true
        case (.error(let lhsMessage), .error(let rhsMessage)):
            return lhsMessage == rhsMessage
        default:
            return false
        }
    }
}

#if canImport(UIKit)
import UIKit
import FirebaseCore
#endif

File: cloud-mobile-integration/firebase/flutter/firebase_auth_manager.dart
import 'package:firebase_auth/firebase_auth.dart';
import 'package:flutter/foundation.dart';
import 'package:google_sign_in/google_sign_in.dart';
import 'package:sign_in_with_apple/sign_in_with_apple.dart';

/// Firebase Authentication Manager for Flutter
/// Handles user authentication with Firebase Auth
class FirebaseAuthManager {
  static final FirebaseAuthManager _instance = FirebaseAuthManager._internal();
  factory FirebaseAuthManager() => _instance;
  FirebaseAuthManager._internal();

  final FirebaseAuth _auth = FirebaseAuth.instance;
  final GoogleSignIn _googleSignIn = GoogleSignIn();

  final ValueNotifier<User?> currentUser = ValueNotifier(null);

  /// Initialize auth state listener
  void initialize() {
    _auth.authStateChanges().listen((User? user) {
      currentUser.value = user;
      debugPrint('Auth state changed: ${user?.uid}');
    });

    // Set initial user
    currentUser.value = _auth.currentUser;
  }

  /// Sign up with email and password
  Future<User> signUpWithEmail({
    required String email,
    required String password,
  }) async {
    try {
      final userCredential = await _auth.createUserWithEmailAndPassword(
        email: email,
        password: password,
      );

      return userCredential.user!;
    } on FirebaseAuthException catch (e) {
      debugPrint('Error signing up: ${e.message}');
      rethrow;
    }
  }

  /// Sign in with email and password
  Future<User> signInWithEmail({
    required String email,
    required String password,
  }) async {
    try {
      final userCredential = await _auth.signInWithEmailAndPassword(
        email: email,
        password: password,
      );

      return userCredential.user!;
    } on FirebaseAuthException catch (e) {
      debugPrint('Error signing in: ${e.message}');
      rethrow;
    }
  }

  /// Sign in with Google
  Future<User> signInWithGoogle() async {
    try {
      // Trigger Google Sign-In flow
      final GoogleSignInAccount? googleUser = await _googleSignIn.signIn();

      if (googleUser == null) {
        throw Exception('Google sign in aborted');
      }

      // Obtain auth details
      final GoogleSignInAuthentication googleAuth =
          await googleUser.authentication;

      // Create credential
      final credential = GoogleAuthProvider.credential(
        accessToken: googleAuth.accessToken,
        idToken: googleAuth.idToken,
      );

      // Sign in to Firebase
      final userCredential = await _auth.signInWithCredential(credential);

      return userCredential.user!;
    } catch (e) {
      debugPrint('Error signing in with Google: $e');
      rethrow;
    }
  }

  /// Sign in with Apple
  Future<User> signInWithApple() async {
    try {
      // Request Apple ID credential
      final appleCredential = await SignInWithApple.getAppleIDCredential(
        scopes: [
          AppleIDAuthorizationScopes.email,
          AppleIDAuthorizationScopes.fullName,
        ],
      );

      // Create OAuth credential
      final oAuthProvider = OAuthProvider('apple.com');
      final credential = oAuthProvider.credential(
        idToken: appleCredential.identityToken,
        accessToken: appleCredential.authorizationCode,
      );

      // Sign in to Firebase
      final userCredential = await _auth.signInWithCredential(credential);

      return userCredential.user!;
    } catch (e) {
      debugPrint('Error signing in with Apple: $e');
      rethrow;
    }
  }

  /// Sign in with Facebook
  /// Note: Requires facebook_auth package
  Future<User> signInWithFacebook(String accessToken) async {
    try {
      final credential = FacebookAuthProvider.credential(accessToken);
      final userCredential = await _auth.signInWithCredential(credential);

      return userCredential.user!;
    } on FirebaseAuthException catch (e) {
      debugPrint('Error signing in with Facebook: ${e.message}');
      rethrow;
    }
  }

  /// Sign in with phone number
  Future<void> signInWithPhoneNumber({
    required String phoneNumber,
    required Function(String verificationId) onCodeSent,
    required Function(PhoneAuthCredential credential) onAutoVerify,
    required Function(FirebaseAuthException error) onError,
  }) async {
    await _auth.verifyPhoneNumber(
      phoneNumber: phoneNumber,
      verificationCompleted: onAutoVerify,
      verificationFailed: onError,
      codeSent: (String verificationId, int? resendToken) {
        onCodeSent(verificationId);
      },
      codeAutoRetrievalTimeout: (String verificationId) {
        debugPrint('Auto retrieval timeout');
      },
    );
  }

  /// Verify phone code
  Future<User> verifyPhoneCode({
    required String verificationId,
    required String smsCode,
  }) async {
    try {
      final credential = PhoneAuthProvider.credential(
        verificationId: verificationId,
        smsCode: smsCode,
      );

      final userCredential = await _auth.signInWithCredential(credential);
      return userCredential.user!;
    } on FirebaseAuthException catch (e) {
      debugPrint('Error verifying phone code: ${e.message}');
      rethrow;
    }
  }

  /// Sign in anonymously
  Future<User> signInAnonymously() async {
    try {
      final userCredential = await _auth.signInAnonymously();
      return userCredential.user!;
    } on FirebaseAuthException catch (e) {
      debugPrint('Error signing in anonymously: ${e.message}');
      rethrow;
    }
  }

  /// Sign out
  Future<void> signOut() async {
    try {
      await Future.wait([
        _auth.signOut(),
        _googleSignIn.signOut(),
      ]);
    } catch (e) {
      debugPrint('Error signing out: $e');
      rethrow;
    }
  }

  /// Get current user
  User? getCurrentUser() => _auth.currentUser;

  /// Check if user is authenticated
  bool isAuthenticated() => _auth.currentUser != null;

  /// Send email verification
  Future<void> sendEmailVerification() async {
    try {
      await _auth.currentUser?.sendEmailVerification();
    } on FirebaseAuthException catch (e) {
      debugPrint('Error sending email verification: ${e.message}');
      rethrow;
    }
  }

  /// Send password reset email
  Future<void> sendPasswordResetEmail({
    required String email,
  }) async {
    try {
      await _auth.sendPasswordResetEmail(email: email);
    } on FirebaseAuthException catch (e) {
      debugPrint('Error sending password reset: ${e.message}');
      rethrow;
    }
  }

  /// Update email
  Future<void> updateEmail({
    required String newEmail,
  }) async {
    try {
      await _auth.currentUser?.updateEmail(newEmail);
    } on FirebaseAuthException catch (e) {
      debugPrint('Error updating email: ${e.message}');
      rethrow;
    }
  }

  /// Update password
  Future<void> updatePassword({
    required String newPassword,
  }) async {
    try {
      await _auth.currentUser?.updatePassword(newPassword);
    } on FirebaseAuthException catch (e) {
      debugPrint('Error updating password: ${e.message}');
      rethrow;
    }
  }

  /// Update profile
  Future<void> updateProfile({
    String? displayName,
    String? photoURL,
  }) async {
    try {
      await _auth.currentUser?.updateDisplayName(displayName);
      await _auth.currentUser?.updatePhotoURL(photoURL);
      await _auth.currentUser?.reload();
    } on FirebaseAuthException catch (e) {
      debugPrint('Error updating profile: ${e.message}');
      rethrow;
    }
  }

  /// Delete account
  Future<void> deleteAccount() async {
    try {
      await _auth.currentUser?.delete();
    } on FirebaseAuthException catch (e) {
      debugPrint('Error deleting account: ${e.message}');
      rethrow;
    }
  }

  /// Get ID token
  Future<String?> getIdToken({bool forceRefresh = false}) async {
    try {
      return await _auth.currentUser?.getIdToken(forceRefresh);
    } on FirebaseAuthException catch (e) {
      debugPrint('Error getting ID token: ${e.message}');
      rethrow;
    }
  }

  /// Link email credential
  Future<User> linkEmailCredential({
    required String email,
    required String password,
  }) async {
    try {
      final credential = EmailAuthProvider.credential(
        email: email,
        password: password,
      );

      final userCredential =
          await _auth.currentUser?.linkWithCredential(credential);

      return userCredential!.user!;
    } on FirebaseAuthException catch (e) {
      debugPrint('Error linking email: ${e.message}');
      rethrow;
    }
  }

  /// Re-authenticate user
  Future<void> reauthenticate({
    required String email,
    required String password,
  }) async {
    try {
      final credential = EmailAuthProvider.credential(
        email: email,
        password: password,
      );

      await _auth.currentUser?.reauthenticateWithCredential(credential);
    } on FirebaseAuthException catch (e) {
      debugPrint('Error reauthenticating: ${e.message}');
      rethrow;
    }
  }

  /// Reload user data
  Future<void> reloadUser() async {
    try {
      await _auth.currentUser?.reload();
      currentUser.value = _auth.currentUser;
    } on FirebaseAuthException catch (e) {
      debugPrint('Error reloading user: ${e.message}');
      rethrow;
    }
  }

  /// Check if email is verified
  bool isEmailVerified() => _auth.currentUser?.emailVerified ?? false;

  /// Get user metadata
  UserMetadata? getUserMetadata() => _auth.currentUser?.metadata;
}
